// @rd/e2e-token-fetcher/lib/axios/https-agents.ts
import { logVariables } from '../utils/log-utils';
import { getUrlProxyFromEnv } from '../utils/proxy-utils';
import { client } from './axios-client';
import { HttpsCookieAgent, createCookieAgent } from 'http-cookie-agent';
import https from 'https';

require('dotenv').config();

const tough = require('tough-cookie');

function isHTTPS(protocol: any) {
  return typeof protocol === 'string' ? /^https:?$/i.test(protocol) : false;
}

let httpsAgent: any;
let httpsProxyAgent: any;

let cookieJar: any;

logVariables();

if (process.env.E2E_TOKEN_FETCHER_NOACS) {
  /**
   * As I could not use axios-cookiejar-support v2.x to enable cookies at axios level,
   * I used https agents with cookie support thanks to http-cookie-agent.
   */

  cookieJar = new tough.CookieJar();

  httpsAgent = new HttpsCookieAgent({
    jar: cookieJar,
    keepAlive: true,
    rejectUnauthorized: false,
  });

  /**
   * I needed to cookify keepalive-proxy-agent, but I could not pass options
   * to it because it requires using createCookieAgent with the class definition
   * and not an instance set with options.
   *
   * Since I needed options to handle http proxies on port 80, self-signed
   * certificate and proxy basic authentication, I used a customized version
   * of keepalive-proxy-agent
   */

  const KeepAliveProxyAgent = require('./keepalive-proxy-agent');
  const HttpsProxyAgent = createCookieAgent(KeepAliveProxyAgent);

  httpsProxyAgent = new HttpsProxyAgent({ jar: cookieJar });
} else {
  /**
   * "Standard" mode using axios-cookiejar-support v1.x and
   * stock keepalive-proxy-agent
   */

  httpsAgent = new https.Agent({ keepAlive: true, rejectUnauthorized: false });
  const KeepAliveProxyAgent = require('keepalive-proxy-agent');

  const { proxyUrl } = getUrlProxyFromEnv();

  if (proxyUrl !== undefined) {
    const u: URL = new URL(proxyUrl);
    const defaultPort = isHTTPS(u.protocol) ? '443' : '80';

    /**
     * Although the documentation says to use proxy.hostname and not proxy.host,
     * only the later seems to work.
     */

    const proxyOptions: any = {
      host: u.hostname,
      port: u.port !== '' ? u.port : defaultPort,
    };

    if (u.username !== undefined && u.username !== '') {
      proxyOptions['auth'] = `${u.username}:${u.password}`;
    }

    httpsProxyAgent = new KeepAliveProxyAgent({
      proxy: proxyOptions,
      rejectUnauthorized: false,
    });
  }
}

export function getCookieJar() {
  if (process.env.E2E_TOKEN_FETCHER_NOACS) {
    return cookieJar;
  } else {
    return client.defaults.jar;
  }
}

export function resetCookies() {
  cookieJar = new tough.CookieJar();
  if (process.env.E2E_TOKEN_FETCHER_NOACS) {
    httpsAgent.jar = cookieJar;
    httpsProxyAgent.jar = cookieJar;
  } else {
    client.defaults.jar = cookieJar;
  }
}

export { httpsAgent, httpsProxyAgent };
