// @rd/e2e-token-fetcher/lib/axios/keepalive-proxy-agent.ts
/**
 * Fork of https://github.com/mknj/node-keepalive-proxy-agent
 * that manages proxies running on default http & https ports.
 * 
 * This fork's objectives are to set default options that correspond
 * to our needs, as the class will be instanciated without any option
 * when using http-cookie-agent rather than axios-cookiejar-support.
 */

/**
 * I had to patch keealive-proxy-agent because port 80 for a proxy that
 * talks HTTTP is replaced by '' which makes the connection fail.
 * 
 * I also needed to set rejectUnauthorized to false to avoid self-signed
 * proxy certificates issues @ RNTBCI.
 *
 * Last, I implemented proxy basic authentication for RNTBCI,
 * althouth I am not sure it was necessary.
 **/

import log from "loglevel";

require('dotenv').config();

const https = require("https");
const net = require("net");
const url = require("url");

function isHTTPS(protocol: any) {
  return typeof protocol === "string" ? /^https:?$/i.test(protocol) : false;
}

function getProxy() {
  let u = null;
  // Process environment variables in reverse priority order
  if (process.env.HTTP_PROXY !== undefined) {
    u = new url.URL(process.env.HTTP_PROXY);
  }
  if (process.env.http_proxy !== undefined) {
    u = new url.URL(process.env.http_proxy);
  }
  if (process.env.HTTPS_PROXY !== undefined) {
    u = new url.URL(process.env.HTTPS_PROXY);
  }
  if (process.env.https_proxy !== undefined) {
    u = new url.URL(process.env.https_proxy);
  }
  if (u) {
    const defaultPort = isHTTPS(u.protocol) ? "443" : "80";
    let proxyOptions: any = {
      hostname: u.hostname,
      port: u.port !== "" ? u.port : defaultPort,
    };
    if (u.username !== undefined && u.username !== '') {
      proxyOptions["auth"] = `${u.username}:${u.password}`;
    }
    log.debug(`* Using proxy options: ${JSON.stringify(proxyOptions)}`)
    return proxyOptions;
  } else {
    return undefined;
  }
}

class KeepAliveProxyAgent extends require("https").Agent {
  constructor(options: any) {
    options = options || {};
    if (options.proxy === undefined) {
      options.proxy = getProxy();
    }
    if (options.keepAlive === undefined) {
      options.keepAlive = true;
    }
    if (options.rejectUnauthorized === undefined) {
      options.rejectUnauthorized = false;
    }
    super(options);
  }

  createConnectionHttpsAfterHttp(options: any, cb: any) {
    const proxySocket = net.connect(
      +options.proxy.port,
      options.proxy.hostname || options.proxy.host
    );
    const errorListener = (error: any) => {
      proxySocket.destroy();
      cb(error);
    };
    proxySocket.once("error", errorListener);

    let response = "";
    const dataListener = (data: any) => {
      response += data.toString();
      if (!response.endsWith("\r\n\r\n")) {
        // response not completed yet
        return;
      }
      proxySocket.removeListener("error", errorListener);
      proxySocket.removeListener("data", dataListener);

      const m = response.match(/^HTTP\/1.\d (\d*)/);
      if (m == null || m[1] == null) {
        proxySocket.destroy();
        return cb(new Error(response.trim()));
      } else if (m[1] !== "200") {
        proxySocket.destroy();
        return cb(new Error(m[0]));
      }
      options.socket = proxySocket; // tell super function to use our proxy socket,
      cb(null, super.createConnection(options));
    };
    proxySocket.on("data", dataListener);

    let host = options.hostname;
    if (!host) host = options.host;
    let cmd = "CONNECT " + host + ":" + options.port + " HTTP/1.1\r\n";
    if (options.proxy.auth) {
      // noinspection JSCheckFunctionSignatures
      const auth = Buffer.from(options.proxy.auth).toString("base64");
      cmd += "Proxy-Authorization: Basic " + auth + "\r\n";
    }
    cmd += "\r\n";
    proxySocket.write(cmd);
  }

  createConnection(options: any, cb: any) {
    if (options.proxy) {
      this.createConnectionHttpsAfterHttp(options, cb);
    } else {
      cb(null, super.createConnection(options));
    }
  }
}

module.exports = KeepAliveProxyAgent;
