// @rd/e2e-token-fetcher/lib/common/get-authentication-cookie.ts
import log from 'loglevel';
import { resetCookies } from '../axios';
import { appNetIQAuthenticationWorkflow } from '../netiq/app-netiq-authentication-workflow';
import { appOktaAuthenticationWorkflow } from '../okta/app-okta-authentication-workflow';
import { addCookieToCache, getCookieFromCache } from '../utils/cache-utils';
import { getCookieFromStore } from '../utils/cookies-utils';
import { setLogLevel } from '../utils/log-utils';
import { urlWithTrailingSlash } from '../utils/url-utils';
import { connectToApp } from './connect-to-app';

require('dotenv').config();

export async function getAuthenticationCookie(
  config: Partial<{
    username: string;
    password: string;
    appUrl: string;
    use_cache: boolean;
  }>,
) {
  if (!config.username) throw new Error('Undefined username');
  if (!config.password) throw new Error('Undefined password');
  if (!config.appUrl) throw new Error('Undefined appUrl');

  config.use_cache = config.use_cache || true;

  const { username, password, appUrl, use_cache } = config;

  setLogLevel();

  let key = `${username}@${urlWithTrailingSlash(appUrl)}`;
  let cookie = use_cache && getCookieFromCache(key);

  if (!cookie) {
    resetCookies(); // Make sure axios cookie store is empty

    // Identify IDP and apply corresponding workflow
    const response = await connectToApp(appUrl);
    if (response.status != 302 || !response.headers.location) {
      const msg = `Application at ${appUrl} did not send a redirection.`;
      throw new Error(msg);
    }
    const idp = response.headers.location.replace('https://', '').split('/')[0];
    log.info(`* Detected IDP '${idp}'`);

    switch (idp) {
      case 'idp.renault.com':
      case 'idp2.renault.com':
        await appNetIQAuthenticationWorkflow(
          appUrl,
          username,
          password,
          response,
        );
        break;
      case 'sso.renault.com':
      case 'sso-uat.renault.com':
        await appOktaAuthenticationWorkflow(
          appUrl,
          username,
          password,
          response,
        );
        break;
      default:
        const msg = `Application at ${appUrl} redirected to unexpected IDP: ${idp}.`;
        throw new Error(msg);
    }

    cookie = getCookieFromStore(appUrl, 'mod_auth_openidc_session');
    if (!cookie) {
      const msg = `Cound not find cookie 'mod_auth_openidc_session' for '${appUrl}' in store.`;
      throw new Error(msg);
    }
    addCookieToCache(cookie, key);
  }

  return cookie;
}
