// @rd/xray-connector/lib/cli/get-gatling-report.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGatlingReport = getGatlingReport;
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var config_1 = require("../config");
var features_1 = require("../features");
function getGatlingReport() {
    var mergedReportDirectory = (0, config_1.getMergedReportDirectory)();
    // Create the merged report directory if needed
    if (!fs_1.default.existsSync(mergedReportDirectory)) {
        fs_1.default.mkdirSync(mergedReportDirectory, { recursive: true });
    }
    generateReport(getTests(), (0, features_1.getGatlingStatus)());
}
/**
 * Get Xray tests from features
 * Tests are expected to contain a line with @<JIRA_PROJECT>-<TEST_ID>
 */
function getTests() {
    var featuresFolder = path_1.default.join((0, config_1.getFeaturesDownloadDirectory)(), 'ACTIVE');
    var tests = [];
    if (fs_1.default.existsSync(featuresFolder)) {
        fs_1.default.readdirSync(featuresFolder)
            .filter(function (name) { return name.endsWith('.feature'); })
            .forEach(function (feature) {
            fs_1.default.readFileSync(path_1.default.join(featuresFolder, feature))
                .toString('utf8')
                .split(/\r?\n/)
                .forEach(function (line) {
                var result = new RegExp('@(\\S*) ').exec(line);
                if (result && result.length > 1) {
                    tests.push('@' + result[1].trim());
                }
            });
        });
    }
    return tests;
}
/**
 * Generate merged-output.json
 *
 * A report file is (minimally) expected to be:
 * [
 *   {
 *     "elements": [
 *       {
 *         "id": "xxx",
 *         "name": "xxx",
 *         "description": "xxx",
 *         "type": "scenario",
 *         "keyword": "Scenario",
 *         "steps": [
 *           {
 *             "name": "xxx",
 *             "result": {
 *               "status": "passed" | "failed"
 *             }
 *           }, ...
 *         ],
 *         "keyword": "Given | When | Then ",
 *         "tags": [ { "name": "@xxx-yyyy" } ]
 *       }, ...
 *     ]
 *   }
 * ]
 */
function generateReport(tests, gatlingStatus) {
    var mergedReportDirectory = (0, config_1.getMergedReportDirectory)();
    fs_1.default.writeFileSync(path_1.default.join(mergedReportDirectory, 'merged-output.json'), JSON.stringify([
        {
            elements: tests.map(function (test) {
                return {
                    id: test,
                    name: test,
                    description: '',
                    type: 'scenario',
                    keyword: 'Scenario',
                    steps: [
                        {
                            name: test,
                            result: { status: gatlingStatus.status },
                            keyword: '',
                        },
                    ],
                    tags: [{ name: test }],
                };
            }),
        },
    ]));
}
