// @rd/e2e-token-fetcher/lib/okta/authn-send-passord.ts
import { request } from '../axios';

// Previous version - Stopped working 31/01/23 between 5:20 pm and 8:00 pm
export async function _authnSendPassword(
  password: string,
  stateToken: string,
  oktaEnv: string,
) {
  const authnPasswordBody = {
    password: `${password}`,
    stateToken: `${stateToken}`,
  };
  const authnPasswordUrl = `https://${oktaEnv}/api/v1/authn/factors/password/verify?rememberDevice=false`;
  return request({
    method: 'POST',
    url: authnPasswordUrl,
    data: authnPasswordBody,
  });
}

export async function authnSendPassword(
  password: string,
  stateToken: string,
  oktaEnv: string,
) {
  const authnPasswordBody = {
    credentials: {
      passcode: `${password}`,
    },
    stateHandle: `${stateToken}`,
  };
  const authnPasswordUrl = `https://${oktaEnv}/idp/idx/challenge/answer`;
  let response: any;
  try {
    response = await request({
      method: 'POST',
      url: authnPasswordUrl,
      data: authnPasswordBody,
    });
  } catch (error) {
    const message = (error as Error).message;
    if (message.indexOf('status code 40') != -1) {
      console.log('** Password sending failed with status code 40x. This might mean login/password authentication has not been set for your application.');
    }
    throw error;
  }
  return response;
}
