// @rd/e2e-token-fetcher/dist/axios/https-agents.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.httpsProxyAgent = exports.httpsAgent = void 0;
exports.getCookieJar = getCookieJar;
exports.resetCookies = resetCookies;
var log_utils_1 = require("../utils/log-utils");
var proxy_utils_1 = require("../utils/proxy-utils");
var axios_client_1 = require("./axios-client");
var http_cookie_agent_1 = require("http-cookie-agent");
var https_1 = __importDefault(require("https"));
require('dotenv').config();
var tough = require('tough-cookie');
function isHTTPS(protocol) {
    return typeof protocol === 'string' ? /^https:?$/i.test(protocol) : false;
}
var httpsAgent;
var httpsProxyAgent;
var cookieJar;
(0, log_utils_1.logVariables)();
if (process.env.E2E_TOKEN_FETCHER_NOACS) {
    /**
     * As I could not use axios-cookiejar-support v2.x to enable cookies at axios level,
     * I used https agents with cookie support thanks to http-cookie-agent.
     */
    cookieJar = new tough.CookieJar();
    exports.httpsAgent = httpsAgent = new http_cookie_agent_1.HttpsCookieAgent({
        jar: cookieJar,
        keepAlive: true,
        rejectUnauthorized: false,
    });
    /**
     * I needed to cookify keepalive-proxy-agent, but I could not pass options
     * to it because it requires using createCookieAgent with the class definition
     * and not an instance set with options.
     *
     * Since I needed options to handle http proxies on port 80, self-signed
     * certificate and proxy basic authentication, I used a customized version
     * of keepalive-proxy-agent
     */
    var KeepAliveProxyAgent = require('./keepalive-proxy-agent');
    var HttpsProxyAgent = (0, http_cookie_agent_1.createCookieAgent)(KeepAliveProxyAgent);
    exports.httpsProxyAgent = httpsProxyAgent = new HttpsProxyAgent({ jar: cookieJar });
}
else {
    /**
     * "Standard" mode using axios-cookiejar-support v1.x and
     * stock keepalive-proxy-agent
     */
    exports.httpsAgent = httpsAgent = new https_1.default.Agent({ keepAlive: true, rejectUnauthorized: false });
    var KeepAliveProxyAgent = require('keepalive-proxy-agent');
    var proxyUrl = (0, proxy_utils_1.getUrlProxyFromEnv)().proxyUrl;
    if (proxyUrl !== undefined) {
        var u = new URL(proxyUrl);
        var defaultPort = isHTTPS(u.protocol) ? '443' : '80';
        /**
         * Although the documentation says to use proxy.hostname and not proxy.host,
         * only the later seems to work.
         */
        var proxyOptions = {
            host: u.hostname,
            port: u.port !== '' ? u.port : defaultPort,
        };
        if (u.username !== undefined && u.username !== '') {
            proxyOptions['auth'] = "".concat(u.username, ":").concat(u.password);
        }
        exports.httpsProxyAgent = httpsProxyAgent = new KeepAliveProxyAgent({
            proxy: proxyOptions,
            rejectUnauthorized: false,
        });
    }
}
function getCookieJar() {
    if (process.env.E2E_TOKEN_FETCHER_NOACS) {
        return cookieJar;
    }
    else {
        return axios_client_1.client.defaults.jar;
    }
}
function resetCookies() {
    cookieJar = new tough.CookieJar();
    if (process.env.E2E_TOKEN_FETCHER_NOACS) {
        httpsAgent.jar = cookieJar;
        httpsProxyAgent.jar = cookieJar;
    }
    else {
        axios_client_1.client.defaults.jar = cookieJar;
    }
}
