// @rd/e2e-token-fetcher/dist/axios/keepalive-proxy-agent.js
"use strict";
/**
 * Fork of https://github.com/mknj/node-keepalive-proxy-agent
 * that manages proxies running on default http & https ports.
 *
 * This fork's objectives are to set default options that correspond
 * to our needs, as the class will be instanciated without any option
 * when using http-cookie-agent rather than axios-cookiejar-support.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * I had to patch keealive-proxy-agent because port 80 for a proxy that
 * talks HTTTP is replaced by '' which makes the connection fail.
 *
 * I also needed to set rejectUnauthorized to false to avoid self-signed
 * proxy certificates issues @ RNTBCI.
 *
 * Last, I implemented proxy basic authentication for RNTBCI,
 * althouth I am not sure it was necessary.
 **/
var loglevel_1 = __importDefault(require("loglevel"));
require('dotenv').config();
var https = require("https");
var net = require("net");
var url = require("url");
function isHTTPS(protocol) {
    return typeof protocol === "string" ? /^https:?$/i.test(protocol) : false;
}
function getProxy() {
    var u = null;
    // Process environment variables in reverse priority order
    if (process.env.HTTP_PROXY !== undefined) {
        u = new url.URL(process.env.HTTP_PROXY);
    }
    if (process.env.http_proxy !== undefined) {
        u = new url.URL(process.env.http_proxy);
    }
    if (process.env.HTTPS_PROXY !== undefined) {
        u = new url.URL(process.env.HTTPS_PROXY);
    }
    if (process.env.https_proxy !== undefined) {
        u = new url.URL(process.env.https_proxy);
    }
    if (u) {
        var defaultPort = isHTTPS(u.protocol) ? "443" : "80";
        var proxyOptions = {
            hostname: u.hostname,
            port: u.port !== "" ? u.port : defaultPort,
        };
        if (u.username !== undefined && u.username !== '') {
            proxyOptions["auth"] = "".concat(u.username, ":").concat(u.password);
        }
        loglevel_1.default.debug("* Using proxy options: ".concat(JSON.stringify(proxyOptions)));
        return proxyOptions;
    }
    else {
        return undefined;
    }
}
var KeepAliveProxyAgent = /** @class */ (function (_super) {
    __extends(KeepAliveProxyAgent, _super);
    function KeepAliveProxyAgent(options) {
        options = options || {};
        if (options.proxy === undefined) {
            options.proxy = getProxy();
        }
        if (options.keepAlive === undefined) {
            options.keepAlive = true;
        }
        if (options.rejectUnauthorized === undefined) {
            options.rejectUnauthorized = false;
        }
        return _super.call(this, options) || this;
    }
    KeepAliveProxyAgent.prototype.createConnectionHttpsAfterHttp = function (options, cb) {
        var _this = this;
        var proxySocket = net.connect(+options.proxy.port, options.proxy.hostname || options.proxy.host);
        var errorListener = function (error) {
            proxySocket.destroy();
            cb(error);
        };
        proxySocket.once("error", errorListener);
        var response = "";
        var dataListener = function (data) {
            response += data.toString();
            if (!response.endsWith("\r\n\r\n")) {
                // response not completed yet
                return;
            }
            proxySocket.removeListener("error", errorListener);
            proxySocket.removeListener("data", dataListener);
            var m = response.match(/^HTTP\/1.\d (\d*)/);
            if (m == null || m[1] == null) {
                proxySocket.destroy();
                return cb(new Error(response.trim()));
            }
            else if (m[1] !== "200") {
                proxySocket.destroy();
                return cb(new Error(m[0]));
            }
            options.socket = proxySocket; // tell super function to use our proxy socket,
            cb(null, _super.prototype.createConnection.call(_this, options));
        };
        proxySocket.on("data", dataListener);
        var host = options.hostname;
        if (!host)
            host = options.host;
        var cmd = "CONNECT " + host + ":" + options.port + " HTTP/1.1\r\n";
        if (options.proxy.auth) {
            // noinspection JSCheckFunctionSignatures
            var auth = Buffer.from(options.proxy.auth).toString("base64");
            cmd += "Proxy-Authorization: Basic " + auth + "\r\n";
        }
        cmd += "\r\n";
        proxySocket.write(cmd);
    };
    KeepAliveProxyAgent.prototype.createConnection = function (options, cb) {
        if (options.proxy) {
            this.createConnectionHttpsAfterHttp(options, cb);
        }
        else {
            cb(null, _super.prototype.createConnection.call(this, options));
        }
    };
    return KeepAliveProxyAgent;
}(require("https").Agent));
module.exports = KeepAliveProxyAgent;
