// @rd/xray-connector/lib/model/test-plan.js
"use strict";
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestPlan = void 0;
var _1 = require(".");
var TestPlan = /** @class */ (function () {
    function TestPlan() {
        this.successfulRequestsPercent = 100;
        this.responseTimeMin = undefined;
        this.responseTimeMax = undefined;
        this.responseTimeMean = undefined;
        this.responseTimeStdDev = undefined;
        this.responseTimePercentile50th = undefined;
        this.responseTimePercentile75th = undefined;
        this.responseTimePercentile95th = undefined;
        this.responseTimePercentile99th = undefined;
        this.requestsPerSecond = undefined;
        this.simulations = [];
    }
    TestPlan.fromXray = function (object) {
        var testPlan = new TestPlan();
        var lines = (object.description || '').split(/[\r\n]+/);
        TestPlan.parseDescriptionLine(lines, testPlan);
        return testPlan;
    };
    TestPlan.parseDescriptionLine = function (lines, testPlan) {
        var result;
        // prettier-ignore
        lines.forEach(function (line) {
            line = line.trim();
            result = new RegExp('^ASSERT_RESPONSE_TIME_MIN=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.responseTimeMin = parseInt(result[1], 10);
            }
            result = new RegExp('^ASSERT_RESPONSE_TIME_MAX=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.responseTimeMax = parseInt(result[1], 10);
            }
            result = new RegExp('^ASSERT_RESPONSE_TIME_MEAN=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.responseTimeMean = parseInt(result[1], 10);
            }
            result = new RegExp('^ASSERT_RESPONSE_TIME_STD_DEV=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.responseTimeStdDev = parseInt(result[1], 10);
            }
            result = new RegExp('^ASSERT_RESPONSE_TIME_PERCENTILE_50TH=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.responseTimePercentile50th = parseInt(result[1], 10);
            }
            result = new RegExp('^ASSERT_RESPONSE_TIME_PERCENTILE_75TH=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.responseTimePercentile75th = parseInt(result[1], 10);
            }
            result = new RegExp('^ASSERT_RESPONSE_TIME_PERCENTILE_95TH=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.responseTimePercentile95th = parseInt(result[1], 10);
            }
            result = new RegExp('^ASSERT_RESPONSE_TIME_PERCENTILE_99TH=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.responseTimePercentile99th = parseInt(result[1], 10);
            }
            result = new RegExp('^ASSERT_REQUESTS_PER_SECOND=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                testPlan.requestsPerSecond = parseInt(result[1], 10);
            }
            result = new RegExp('^TAG=(\\S*)$').exec(line);
            if (result && result.length > 1) {
                var simulation = new _1.TestSimulation();
                simulation.tag = result[1];
                testPlan.simulations.push(simulation);
            }
            result = new RegExp('^INJECTION_PROFILE=(\\S*)$').exec(line);
            if (result && result.length > 1) {
                var simulation = TestPlan.getSimulation(testPlan);
                var step = new _1.TestSimulationStep();
                step.injectionProfile = result[1];
                simulation.injectionSteps.push(step);
            }
            result = new RegExp('^USERS=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                var simulation = TestPlan.getSimulation(testPlan);
                var step = TestPlan.getStep(simulation);
                step.users = parseInt(result[1], 10);
            }
            result = new RegExp('^DURATION=(\\d*)$').exec(line);
            if (result && result.length > 1) {
                var simulation = TestPlan.getSimulation(testPlan);
                var step = TestPlan.getStep(simulation);
                step.duration = parseInt(result[1], 10);
            }
        });
    };
    TestPlan.getSimulation = function (testPlan) {
        if (testPlan.simulations.length > 0) {
            return testPlan.simulations[testPlan.simulations.length - 1];
        }
        return new _1.TestSimulation();
    };
    TestPlan.getStep = function (simulation) {
        if (simulation && simulation.injectionSteps.length > 0) {
            return simulation.injectionSteps[simulation.injectionSteps.length - 1];
        }
        return new _1.TestSimulationStep();
    };
    return TestPlan;
}());
exports.TestPlan = TestPlan;
