// @rd/xray-connector/lib/interface/build-execution-info.js
"use strict";
// https://wanago.io/2019/03/18/node-js-typescript-6-sending-http-requests-understanding-multipart-form-data/
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildExecutionInfo = buildExecutionInfo;
var config_1 = require("../config");
function buildExecutionInfo(summary, description, testPlanKey, labels, fixVersions, testEnvironments) {
    var fieldsIds = (0, config_1.getXrayFieldsIds)();
    var testExecutionTestEnvionmentsFieldId = 'customfield_' + fieldsIds.testExecutionTestEnvironmentsFieldId;
    var testExecutionTestPlansFieldId = 'customfield_' + fieldsIds.testExecutionTestPlansFieldId;
    var projectKey = testPlanKey.split('-')[0];
    // prettier-ignore
    var info = {
        fields: {
            summary: summary,
            description: description,
            labels: labels,
            project: {
                key: projectKey,
            },
            issuetype: {
                id: '' + fieldsIds.testExecutionIssueTypeId,
            },
        },
    };
    info.fields[testExecutionTestEnvionmentsFieldId] = [];
    info.fields[testExecutionTestPlansFieldId] = [testPlanKey];
    // Add environments if provided
    if (testEnvironments && testEnvironments.length !== 0) {
        info.fields[testExecutionTestEnvionmentsFieldId] = testEnvironments;
    }
    // Add fix version if provided
    if (fixVersions && fixVersions.length !== 0) {
        info.fields.fixVersions = fixVersions;
    }
    return info;
}
