// @rd/xray-connector/lib/interface/create-test-plan-execution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTestPlanExecution = createTestPlanExecution;
var build_execution_info_1 = require("./build-execution-info");
var jira_client_1 = require("./jira-client");
// https://confluence.xpand-it.com/display/public/XRAY/Import+Execution+Results+-+REST
/**
 * Creates an execution for test plan in Xray
 */
function createTestPlanExecution(summary, description, results, testPlanKey, testLabels, testFixVersions, testEnvironment) {
    var jira = (0, jira_client_1.getJiraClient)();
    var header = jira.makeRequestHeader(jira.makeXrayUri({
        pathname: '/import/execution/cucumber/multipart',
    }));
    header.method = 'POST';
    var info = JSON.stringify((0, build_execution_info_1.buildExecutionInfo)(summary, description, testPlanKey, testLabels ? testLabels.split(',') : [], testFixVersions ? testFixVersions.split(',') : [], testEnvironment ? [testEnvironment] : undefined));
    var result = {
        value: JSON.stringify(results),
        options: {
            filename: 'merged-output.json',
            contentType: 'application/json',
        },
    };
    var formData = {
        info: info,
        result: result,
    };
    header.formData = formData;
    return jira.doRequest(header);
}
