// @rd/xray-connector/lib/interface/jira-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getJiraClient = getJiraClient;
exports.destroyJiraClient = destroyJiraClient;
var jira_client_xray_1 = __importDefault(require("jira-client-xray"));
var config_1 = require("../config");
var jira;
function initJiraClient() {
    var credentials = (0, config_1.getXrayCredentials)();
    var jiraClientOptions = {
        strictSSL: true,
        protocol: 'https',
        host: (0, config_1.getXrayHost)(),
        base: '',
        xrayVersion: '1.0',
    };
    if (credentials.token) {
        jiraClientOptions.bearer = credentials.token;
    }
    else {
        jiraClientOptions.username = credentials.username;
        jiraClientOptions.password = credentials.password;
    }
    return new jira_client_xray_1.default(jiraClientOptions);
}
function getJiraClient() {
    return jira || initJiraClient();
}
function destroyJiraClient() {
    jira = undefined;
}
