// @rd/e2e-token-fetcher/lib/netiq/get-netiq-access-token.ts
import { request } from '../axios';
import { Token } from '../types';
import { addTokenToCache, getTokenFromCache } from '../utils/cache-utils';
import { setLogLevel } from '../utils/log-utils';
import log from 'loglevel';

require('dotenv').config();

export async function getNetIQAccessToken(
  config: Partial<{
    username: string;
    password: string;
    client_id: string;
    client_secret: string;
    grant_type: string;
    scope: string;
    idp: string;
    use_cache: boolean;
  }>,
) {
  if (!config.username) throw new Error('Undefined username');
  if (!config.password) throw new Error('Undefined password');
  if (!config.client_id) throw new Error('Undefined client_id');
  if (!config.client_secret) throw new Error('Undefined username');

  config.grant_type = config.grant_type || 'password';
  config.scope = config.scope || 'arca';
  config.idp = config.idp || 'idp2';
  config.use_cache = config.use_cache || true;

  if (['idp', 'idp2'].indexOf(config.idp) === -1)
    throw new Error(`Unkonwn IDP: ${config.idp}. Expected 'idp' or 'idp2'.`);

  const {
    username,
    password,
    client_id,
    client_secret,
    grant_type,
    scope,
    idp,
    use_cache,
  } = config;

  setLogLevel();

  const msg = `* Getting NetIQ access token for '${username}' for application identified by '${client_id}'`;
  log.info(msg);

  let key = `${grant_type}:${username}@${client_id}:${scope}`;
  let token = use_cache && getTokenFromCache(key);

  if (!token) {
    const idpTokenURL = `https://${idp}.renault.com/nidp/oauth/nam/token`;
    const params = [
      `grant_type=${grant_type}`,
      `username=${username}`,
      `password=${password}`,
      `client_id=${client_id}`,
      `client_secret=${client_secret}`,
      `scope=${scope}`,
    ];
    const url = `${idpTokenURL}?` + params.join('&');
    let response: any = await request({ method: 'POST', url });
    token = response.data as Token;
    addTokenToCache(token, key);
  }

  return token;
}
