// @rd/e2e-token-fetcher/lib/utils/cache-utils.ts
import { Cookie, Token } from '../types';
import log from 'loglevel';

const DEFAULT_COOKIE_EXPIRATION = 300; // 5 minutes

const cookiesCache: any[] = [];
const tokensCache: any[] = [];

export function getCookieFromCache(key: string): Cookie | undefined {
  let cacheEntry = cookiesCache.find((entry) => entry.key == key);
  if (cacheEntry) {
    // Expire after 300 seconds as a default
    const expirationDate = new Date(
      cacheEntry.insertionDate.getTime() +
        (Number(process.env.E2E_TOKEN_FETCHER_COOKIE_EXPIRATION) ||
          DEFAULT_COOKIE_EXPIRATION) *
          1000,
    );
    if (expirationDate > new Date()) {
      log.info('* Getting authentication cookie from cache');
      return cacheEntry.cookie;
    } else {
      log.info('* Expired cookie found in cache, getting a new one');
      cacheEntry.key += '_expired';
      return undefined;
    }
  } else {
    return undefined;
  }
}

export function addCookieToCache(cookie: Cookie, key: string): void {
  let cacheEntry = cookiesCache.find((entry) => entry.key == key);
  if (cacheEntry) {
    cacheEntry.cookie = cookie;
    cacheEntry.insertionDate = new Date();
  } else {
    log.info('* Adding authentication cookie to cache');
    cookiesCache.push({ key, cookie, insertionDate: new Date() });
  }
}

// Only used for unit testing
export function expireCookie(key: string): void {
  let cacheEntry = cookiesCache.find((entry) => entry.key == key);
  if (cacheEntry) {
    cacheEntry.insertionDate = new Date(0);
  }
}

export function getTokenFromCache(key: string): Token | undefined {
  let cacheEntry = tokensCache.find((entry) => entry.key == key);
  if (cacheEntry) {
    const expirationDate = new Date(
      cacheEntry.insertionDate.getTime() + cacheEntry.token.expires_in * 1000,
    );
    if (expirationDate > new Date()) {
      log.info('* Getting access token from cache');
      return cacheEntry.token;
    } else {
      log.info('* Expired access token found in cache, getting a new one');
      cacheEntry.key += '_expired';
      return undefined;
    }
  } else {
    return undefined;
  }
}

export function addTokenToCache(token: Token, key: string): void {
  let cacheEntry = tokensCache.find((entry) => entry.key == key);
  if (cacheEntry) {
    cacheEntry.token = token;
    cacheEntry.insertionDate = new Date();
  } else {
    log.info('* Adding access token to cache');
    tokensCache.push({ key, token, insertionDate: new Date() });
  }
}

// Only used for unit testing
export function expireToken(key: string): void {
  let cacheEntry = tokensCache.find((entry) => entry.key == key);
  if (cacheEntry) {
    cacheEntry.token.expires_in = 0;
  }
}
