// @rd/e2e-token-fetcher/dist/utils/cache-utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCookieFromCache = getCookieFromCache;
exports.addCookieToCache = addCookieToCache;
exports.expireCookie = expireCookie;
exports.getTokenFromCache = getTokenFromCache;
exports.addTokenToCache = addTokenToCache;
exports.expireToken = expireToken;
var loglevel_1 = __importDefault(require("loglevel"));
var DEFAULT_COOKIE_EXPIRATION = 300; // 5 minutes
var cookiesCache = [];
var tokensCache = [];
function getCookieFromCache(key) {
    var cacheEntry = cookiesCache.find(function (entry) { return entry.key == key; });
    if (cacheEntry) {
        // Expire after 300 seconds as a default
        var expirationDate = new Date(cacheEntry.insertionDate.getTime() +
            (Number(process.env.E2E_TOKEN_FETCHER_COOKIE_EXPIRATION) ||
                DEFAULT_COOKIE_EXPIRATION) *
                1000);
        if (expirationDate > new Date()) {
            loglevel_1.default.info('* Getting authentication cookie from cache');
            return cacheEntry.cookie;
        }
        else {
            loglevel_1.default.info('* Expired cookie found in cache, getting a new one');
            cacheEntry.key += '_expired';
            return undefined;
        }
    }
    else {
        return undefined;
    }
}
function addCookieToCache(cookie, key) {
    var cacheEntry = cookiesCache.find(function (entry) { return entry.key == key; });
    if (cacheEntry) {
        cacheEntry.cookie = cookie;
        cacheEntry.insertionDate = new Date();
    }
    else {
        loglevel_1.default.info('* Adding authentication cookie to cache');
        cookiesCache.push({ key: key, cookie: cookie, insertionDate: new Date() });
    }
}
// Only used for unit testing
function expireCookie(key) {
    var cacheEntry = cookiesCache.find(function (entry) { return entry.key == key; });
    if (cacheEntry) {
        cacheEntry.insertionDate = new Date(0);
    }
}
function getTokenFromCache(key) {
    var cacheEntry = tokensCache.find(function (entry) { return entry.key == key; });
    if (cacheEntry) {
        var expirationDate = new Date(cacheEntry.insertionDate.getTime() + cacheEntry.token.expires_in * 1000);
        if (expirationDate > new Date()) {
            loglevel_1.default.info('* Getting access token from cache');
            return cacheEntry.token;
        }
        else {
            loglevel_1.default.info('* Expired access token found in cache, getting a new one');
            cacheEntry.key += '_expired';
            return undefined;
        }
    }
    else {
        return undefined;
    }
}
function addTokenToCache(token, key) {
    var cacheEntry = tokensCache.find(function (entry) { return entry.key == key; });
    if (cacheEntry) {
        cacheEntry.token = token;
        cacheEntry.insertionDate = new Date();
    }
    else {
        loglevel_1.default.info('* Adding access token to cache');
        tokensCache.push({ key: key, token: token, insertionDate: new Date() });
    }
}
// Only used for unit testing
function expireToken(key) {
    var cacheEntry = tokensCache.find(function (entry) { return entry.key == key; });
    if (cacheEntry) {
        cacheEntry.token.expires_in = 0;
    }
}
