// @rd/xray-connector/lib/features/get-gatling-status.js
"use strict";
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-return */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatlingStatus = exports.GatlingAssertion = void 0;
exports.getGatlingStatus = getGatlingStatus;
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var config_1 = require("../config");
var GatlingAssertion = /** @class */ (function () {
    function GatlingAssertion(status, message) {
        this.status = status;
        this.message = message;
    }
    return GatlingAssertion;
}());
exports.GatlingAssertion = GatlingAssertion;
var GatlingStatus = /** @class */ (function () {
    function GatlingStatus(status, assertions) {
        this.status = status;
        this.assertions = assertions;
    }
    return GatlingStatus;
}());
exports.GatlingStatus = GatlingStatus;
/**
 * Parse global status from assertions.json
 *
 * assertions.json is expected to be:
 *
 * {
 *   "assertions": [
 *     {
 *       "message": "xxx",
 *       "result": true | false
 *     }, ...
 *   ]
 * }
 */
function parseIntFromString(string, prefix) {
    return parseInt(string.substring(prefix.length + 1), 10);
}
function getGatlingStatus() {
    var prefix = (0, config_1.getGatlingSimulation)().toLowerCase();
    var folder = fs_1.default
        .readdirSync((0, config_1.getGatlingReportsDirectory)())
        .filter(function (name) { return name.startsWith(prefix); })
        .sort(function (a, b) { return parseIntFromString(b, prefix) - parseIntFromString(a, prefix); })
        .shift();
    if (!folder) {
        return new GatlingStatus('aborted', []);
    }
    var assertionFile = path_1.default.join((0, config_1.getGatlingReportsDirectory)(), folder, 'js', 'assertions.json');
    if (!fs_1.default.existsSync(assertionFile)) {
        return new GatlingStatus('aborted', []);
    }
    var assertionJson = JSON.parse(fs_1.default.readFileSync(assertionFile).toString('utf8'));
    var assertions = assertionJson['assertions'] || [];
    var reducedStatus = assertions.reduce(function (status, assertion) { return status && assertion['result']; }, true);
    return new GatlingStatus(reducedStatus ? 'passed' : 'failed', assertions.map(function (assertion) { return new GatlingAssertion(assertion['result'], assertion['message']); }));
}
